packages/tasit-action/src/contract/Action.js

```diff
@@ -6,32 +6,67 @@ import ConfigLoader from "../ConfigLoader";
 //  and/or MetaTxAction subclasses
 export class Action extends Subscription {
   #provider;
+  #signer;
+  #rawTx;
   #tx;
   #txConfirmations;
   #timeout;
   #lastConfirmationTime;
 
-  constructor(txPromise, provider) {
+  constructor(rawTx, provider, signer) {
     // Provider implements EventEmitter API and it's enough
     //  to handle with transactions events
     super(provider);
 
     const { events } = ConfigLoader.getConfig();
     const { timeout } = events;
 
-    this.#tx = txPromise.then(
+    this.#rawTx = rawTx;
+    this.#signer = signer;
+    this.#timeout = timeout;
+    this.#provider = provider;
+    this.#txConfirmations = 0;
+  }
+
+  _toRaw = async () => {
+    return await this.#rawTx;
```

Comment:
Can we handle whether to await it or not upstream of this function somehow? Like only call _toRaw once we know the promise has resolved? Or only use this in `tasit-contract-based-account`?

Reply:
1. I need to think about it, not sure.
2. For now, we are using that function only on `tasit-contract-based-account` GnosisSafe class.

Reply:
1. Okay. Keep me posted. My intuition tells me that move would make sense, but I haven't thought it through fully.
2. Yep, I thought so. That's part of my rough reasoning for this.

Reply:
I've made a change on this PR https://github.com/tasitlabs/TasitSDK/pull/360
On these changes, we are assuming that the `Action` will expect a sync `rawTx` on the constructor but the `signAndSend` function stills using `await` to resolve that. It isn't 100% solution, though.
Another path is to force `rawTx` to be sync (removing the `await` inside of the `signAndSend` but that'll make `GnosisSafe.executeTransaction` to be async (on the way that is written for now).

Reply:
I don't have time to think this through fully right now - feel free to go with whatever you thinks makes the most sense for now, and I'll circle back to this soon

Reply:
The approach you're using in the latest version of the code works for me.

Just to make sure I understand, why would removing the `await` inside `signAndSend` mean that `GnosisSafe.executeTransaction` would be async?

Reply:
If we remove `await` from the `signAndSend` the `Action` class won't accept `promise-of-rawTx` as a constructor argument.
That way is how I've found to keep `GnosisSafe` action-create-functions as sync (as `Contract` does). ([ref](https://github.com/tasitlabs/TasitSDK/blob/4111170669afb4447cb95315f45fd4ecd1fa02ec/packages/tasit-contract-based-account/src/GnosisSafe.js#L148))
If `Action` stop to accept `txRaw` as a promise, I'm not seeing a way to keep `GnosisSafe.executeTransaction` as sync right now.

Reply:
Ah right, because `GnosisSafe.executeTransaction` is an action-builder, not something that sends the action. I forgot that when reading it - it makes more sense now.

Reply:
Or wait, maybe what I just said is wrong

Reply:
I guess what I'm asking is this: At a high level, what is it about `GnosisSafe.executeTransaction` that prevents it from having a rawTx ready to go right away?

Reply:
> I guess what I'm asking is this: At a high level, what is it about `GnosisSafe.executeTransaction` that prevents it from having a rawTx ready to go right away?

The `[#prepareTransaction](https://github.com/tasitlabs/TasitSDK/blob/4111170669afb4447cb95315f45fd4ecd1fa02ec/packages/tasit-contract-based-account/src/GnosisSafe.js#L157)`.
That function has some async calls needed to create a `safe` transaction (e.g.: GnosisSafeContract.nonce (L199) and GnosisSafeContract.getTransactionHash (L201)).
These async calls are nested and needed to create the transaction that we want to be executed by the `GnosisSafeContract`.
My intention is to have the `GnosisSafe` encapsulating all that is possible, transparent to the user.

Reply:
transparent to the user or opaque to the user?

Reply:
:man_facepalming: I don't know why I've used the "transparent" word. Opaque was what I've meant.